url542ae48576_tonyheadford_m68k_tgz-pJ8-m68k_MonitorJ8/src/m68k/cpu/instructions/MOVEM.java

```diff
@@ -0,0 +1,692 @@
+package m68k.cpu.instructions;
+
+import m68k.cpu.*;
+
+/*
+//  M68k - Java Amiga MachineCore
+//  Copyright (c) 2008-2010, Tony Headford
+//  All rights reserved.
+//
+//  Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+//  following conditions are met:
+//
+//    o  Redistributions of source code must retain the above copyright notice, this list of conditions and the
+//       following disclaimer.
+//    o  Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+//       following disclaimer in the documentation and/or other materials provided with the distribution.
+//    o  Neither the name of the M68k Project nor the names of its contributors may be used to endorse or promote
+//       products derived from this software without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+//  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+*/
+
+// when the ea is specified by a pre-decreasing mode (eg movem.l range,-(an) , then, if the addressing register
+// itself is part of the range and for a 68000 and 68010, the value stored is the INITIAL value of the addressing reg.
+// For a 68020 and up, the value written is the initial value minus the size (word or long) of the operation.
+
+
+public class MOVEM implements InstructionHandler
+{
+	protected final Cpu cpu;
+
+	private static final int[] M2R_Timing = {
+			0, 0, 12, 12, 0, 16, 18, 16, 20, 16, 18
+	};
+	private static final int[] R2M_Timing = {
+			0, 0, 8, 0, 8, 12, 14, 12, 16
+	};
+
+	public MOVEM(Cpu cpu)
+	{
+		this.cpu = cpu;
+	}
+
+	public final void register(InstructionSet is)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 15. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 57-101, lines 104-148 into separate methods.